Prototype for binary serializer to a file stream

hcho3 opened the pull request:
Closes #265 

TODOs
- [x] Expose a C API to (de)serialize to a file stream
- [x] Write a small utility script that lets users convert a scikit-learn pickle into a Treelite binary serialized file
- [x] Write an end-to-end test

hcho3 committed:
Add version.h header

hcho3 committed:
Enforce exact version match when deserializing

hcho3 committed:
Prototype for a file stream serializer

hcho3 committed:
Install version.h when installing Treelite

hcho3 committed:
Use binary mode for I/O (makes difference on Windows)

hcho3 committed:
Add comment for PyBufferFrame

hcho3 committed:
Rename PyBufferFrameWithManagedBuffer -> PyBufferFrameWithManagedBuffers

hcho3 committed:
Reduce number of copies from 2 to 1

hcho3 committed:
Remove unused function CopyFrom()

hcho3 committed:
Further simplify Model::Deserialize()

hcho3 committed:
Fix some memory leak

hcho3 committed:
Eliminate memory leak completely

hcho3 committed:
Hide assume_ownership from public methods, to prevent confusion

hcho3 committed:
Update an outdated comment

hcho3 committed:
Add C API interface for serializer

hcho3 committed:
Perform round-trip serialization twice, to ensure Serialize() is non-destructive

hcho3 committed:
Merge remote-tracking branch 'origin/mainline' into fast_binary_serializer

hcho3 committed:
Add serialize / deserialize methods in Python

hcho3 committed:
Test round-trip serialization in Python

hcho3 commented:
Still need to write a CLI wrapper script for the serializer, but functionality-wise this is ready for review.

hcho3 committed:
Fix style

wphicks commented on include/treelite/tree_impl.h:
```diff
@@ -46,6 +46,69 @@ inline std::string GetString<double>(double x) {
 
+inline void
+PyBufferFrame::Serialize(FILE* dest_fp) const {
+  auto write_to_file = [](const void* buffer, size_t size, size_t count, FILE* fp) {
```
It would be preferable to use `std::size_t` here and include `<cstddef>`, but if that is not possible, we at least need to include `<stddef.h>`

hcho3 replied:
Done

wphicks commented on include/treelite/tree.h:
```diff
@@ -35,13 +36,26 @@ float stof(const std::string& value, size_t* pos);
   size_t nitem;
+
+  // Serialize a frame to a file stream
+  inline void Serialize(FILE* dest_fp) const;
```
Can you remind me why we need to mark this as inline? I remember thinking about this before and concluding that ODR required it for functions in here, but I'm not immediately seeing the ODR justification when I'm looking at it this time.

hcho3 replied:
The body of this function is defined in `tree_impl.h`.

hcho3 replied:
I moved `Serialize()` to its own source file, so it no longer has the `inline` designation.

wphicks commented on include/treelite/tree_impl.h:
```diff
@@ -418,10 +492,14 @@ inline void InitScalarFromPyBuffer(TaskType* scalar, PyBufferFrame buffer) {
   }
   T* t = static_cast<T*>(buffer.buf);
   *scalar = static_cast<TaskType>(*t);
+  if (assume_ownership) {
```
We're doing quite a bit of manual ownership tracking throughout. Would it be possible to use built-in ownership semantics? I.e. If we want to transfer ownership, move a `unique_ptr`, if we want to give access without sharing, provide a `weak_ptr`, etc.?

hcho3 replied:
I redesigned the serialization logic so that the file serializer doesn't depend at all on `PyBufferFrame`, so we don't need to do manual ownership tracking.

wphicks commented on include/treelite/tree_impl.h:
```diff
@@ -705,26 +788,71 @@ Model::Dispatch(Func func) const {
 }
 
+inline std::unique_ptr<Model>
+Model::CreateFromPyBuffer(std::vector<PyBufferFrame> frames) {
```
Is the copy necessary here? Is there a scenario in which someone would call `CreateFromPyBuffer(frames)` and then continue to use those frames later? Could we just accept `std::vector<PyBufferFrame>&&` instead?

hcho3 replied:
> Is there a scenario in which someone would call `CreateFromPyBuffer(frames)` and then continue to use those frames later?

Yes, the `frames` can originate from outside. For example, in MNMG setting, each worker process receives a series of frames that represent a cuML RF model. The frames are held by the Python interpreter, and our C++ code has no control over when they will be used or destroyed.

wphicks reviewed (commented):
This looks really great! The only thing I'm squeamish about is all of the manual memory management we're doing throughout. If possible, I'd love to see us use a little more RAII and indicate ownership through owning/non-owning pointers where needed, but maybe there's some reason for not doing that that I'm not aware of in this case?

hcho3 commented:
> The only thing I'm squeamish about is all of the manual memory management we're doing throughout. If possible, I'd love to see us use a little more RAII and indicate ownership through owning/non-owning pointers where needed

The limitation is `PyBufferFrame` struct, which contains two raw pointers `buf` and `format`. These pointers refer to buffers that might originate from other libraries, e.g. the Python interpreter. The Python buffer protocol ([PEP 3118](https://www.python.org/dev/peps/pep-3118/)) was designed to allow multiple libraries to exchange raw buffers.

We could avoid using `PyBufferFrame` entirely for this PR, but then we'd need to duplicate all the serialization logic that's available in `GetPyBuffer()` and `CreateFromPyBuffer()`. The logic was originally written to support in-memory serialization, but it's also useful for serializing to the disk.

hcho3 committed:
size_t -> std::size_t

hcho3 committed:
Add a CLI script to convert model files into checkpoints

hcho3 commented:
I added a CLI script.

Usage:
```
python -m treelite.serialize --input-model ./model.pkl --input-model-type sklearn_pkl  \
    --output-checkpoint saved.tl
```

Help:
```
usage: __main__.py [-h] --input-model INPUT_MODEL --input-model-type {sklearn_pkl,xgboost,xgboost_json,lightgbm}
                   --output-checkpoint OUTPUT_CHECKPOINT

CLI script to convert a model file to a Treelite binary checkpoint

optional arguments:
  -h, --help            show this help message and exit
  --input-model INPUT_MODEL
                        Path to the tree model file
  --input-model-type {sklearn_pkl,xgboost,xgboost_json,lightgbm}
                        Type of the tree model file
  --output-checkpoint OUTPUT_CHECKPOINT
                        Path to the checkpoint file, using a special binary format for Treelite
```

hcho3 committed:
Fix style

hcho3 committed:
Move serialization functions impl to serializer.cc

Also revise member functions of PyBufferFrame to make them a
free-standing functions instead.

hcho3 committed:
Update copyright year

hcho3 committed:
Don't use PyBufferFrame for file serialization

hcho3 committed:
Zero-fill padding in structs

hcho3 commented:
I redesigned the serialization code so that the file serializer doesn't depend at all on `PyBufferFrame`, so manual ownership tracking is no longer required. I avoided code duplication by factoring out common serialization logic as template functions.

@wphicks Can I get another round of review?

wphicks reviewed (approved):
Looks great! Can't wait to make use of it
